From: Shaohua Li <shli@fb.com>
Date: Mon, 3 Jul 2017 14:34:23 -0700
Subject: [PATCH] MD: fix sleep in atomic
Git-commit: 7184ef8bab0cb865c3cea9dd1a675771145df0af
Patch-mainline: v4.13
References: bsc#1040351

bioset_free() will take a mutex, so can't get called with spinlock hold.

Fix: 5a85071c2cbc(md: use a separate bio_set for synchronous IO.)
Cc: NeilBrown <neilb@suse.com>
Signed-off-by: Shaohua Li <shli@fb.com>
Acked-by: NeilBrown <neilb@suse.com>

---
 drivers/md/md.c |   12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -471,6 +471,8 @@ static void mddev_delayed_delete(struct
 
 static void mddev_put(struct mddev *mddev)
 {
+	struct bio_set *bs = NULL, *sync_bs = NULL;
+
 	if (!atomic_dec_and_lock(&mddev->active, &all_mddevs_lock))
 		return;
 	if (!mddev->raid_disks && list_empty(&mddev->disks) &&
@@ -478,10 +480,8 @@ static void mddev_put(struct mddev *mdde
 		/* Array is not configured at all, and not held active,
 		 * so destroy it */
 		list_del_init(&mddev->all_mddevs);
-		if (mddev->bio_set)
-			bioset_free(mddev->bio_set);
-		if (mddev->sync_set)
-			bioset_free(mddev->sync_set);
+		bs = mddev->bio_set;
+		sync_bs = mddev->sync_set;
 		mddev->bio_set = NULL;
 		mddev->sync_set = NULL;
 		if (mddev->gendisk) {
@@ -496,6 +496,10 @@ static void mddev_put(struct mddev *mdde
 			kfree(mddev);
 	}
 	spin_unlock(&all_mddevs_lock);
+	if (bs)
+		bioset_free(bs);
+	if (sync_bs)
+		bioset_free(sync_bs);
 }
 
 static void md_safemode_timeout(unsigned long data);
